File: 3562.maximum-profit-from-trading-stocks-with-discounts.py
#
# @lc app=leetcode id=3562 lang=python3
#
# [3562] Maximum Profit from Trading Stocks with Discounts
#

# @lc code=start
from typing import List
import math

class Solution:
    def maxProfit(self,
                  n: int,
                  present: List[int],
                  future: List[int],
                  hierarchy: List[List[int]],
                  budget: int) -> int:
        # Build adjacency list
        adj = [[] for _ in range(n+1)]
        for par, ch in hierarchy:
            adj[par].append(ch)
        
        # Precompute half prices
        half_present = [math.floor(p/2) for p in present]
        # Shift indices to 1-based
        present = [0] + present          # index i corresponds to employee i
        future = [0] + future
        half_present = [0] + half_present
        
        # Memoization dictionary
        memo = {}
        
        def dfs(u):
            """Return two lists:
               res[0]: list length budget+1 for state where parent NOT bought
               res[1]: list length budget+1 for state where parent BOUGHT"""
            # Initialize knapsack arrays for combining children
            # cur0 : combined results from children assuming current node NOT bought
            # cur1 : combined results from children assuming current node BOUGHT
            cur0 = [-10**9] * (budget+1)
            cur0[0] = 0
            cur1 = [-10**9] * (budget+1)
            cur1[0] = 0
            
            # Process children
            for v in adj[u]:
                # Recursively compute DP for child
                child_res = dfs(v)
                # For cur0 : current node NOT bought -> children see parent NOT bought
                # Merge cur0 with child_res[0]
                new_cur0 = [-10**9] * (budget+1)
                for i in range(budget+1):
                    if cur0[i] == -10**9:
                        continue
                    for j in range(budget+1-i):
                        if child_res[0][j] == -10**9:
                            continue
                        new_cur0[i+j] = max(new_cur0[i+j], cur0[i] + child_res[0][j])
                cur0 = new_cur0
                
                # For cur1 : current node BOUGHT -> children see parent BOUGHT
                # Merge cur1 with child_res[1]
                new_cur1 = [-10**9] * (budget+1)
                for i in range(budget+1):
                    if cur1[i] == -10**9:
                        continue
                    for j in range(budget+1-i):
                        if child_res[1][j] == -10**9:
                            continue
                        new_cur1[i+j] = max(new_cur1[i+j], cur1[i] + child_res[1][j])
                cur1 = new_cur1
            
            # Now compute results for both states of parent
            res_parent_not_bought = [-10**9] * (budget+1)
            res_parent_bought    = [-10**9] * (budget+1)
            
            # Common options:
            # Option A : do NOT buy current node -> simply use cur0
            # Option B : buy current node -> add self profit after shifting by appropriate cost
            
            # For res_parent_not_bought:
            full_cost = present[u]
            full_profit = future[u] - full_cost
            # Option A
            for c in range(budget+1):
                res_parent_not_bought[c] = max(res_parent_not_bought[c], cur0[c])
            # Option B
            for c in range(full_cost, budget+1):
                prev_profit = cur1[c-full_cost]
                if prev_profit != -10**9:
                    candidate = prev_profit + full_profit
                    res_parent_not_bought[c] = max(res_parent_not_bought[c], candidate)
                    
            # For res_parent_bought:
            disc_cost = half_present[u]
            disc_profit = future[u] - disc_cost
            # Option A again uses cur0
            for c in range(budget+1):
                res_parent_bought[c] = max(res_parent_bought[c], cur0[c])
            # Option B uses discounted self-cost
            for c in range(disc_cost, budget+1):
                prev_profit = cur1[c-disc_cost]
                if prev_profit != -10**9:
                    candidate = prev_profit + disc_profit
                    res_parent_bought[c] = max(res_parent_bought[c], candidate)
                    
            return res_parent_not_bought , res_parent_bought
        
        root_res = dfs(1)[0]   # root has no boss -> treat as parent NOT bought
        return max(root_res)
# @lc code=end

File: 3562.maximum-profit-from-trading-stocks-with-discounts.cpp
#include <bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    int maxProfit(int n, vector<int>& present, vector<int>& future, vector<vector<int>>& hierarchy, int budget) {
        vector<vector<int>> children(n);
        for (auto &e : hierarchy) {
            int u = e[0] - 1, v = e[1] - 1;
            children[u].push_back(v);
        }

        const long long NEG = -(1LL<<60);

        // dp0[u], dp1[u]
        vector<vector<long long>> dp0(n, vector<long long>(budget + 1, NEG));
        vector<vector<long long>> dp1(n, vector<long long>(budget + 1, NEG));

        // postorder traversal
        vector<int> order;
        order.reserve(n);
        vector<int> st = {0};
        vector<int> it(n, 0);
        vector<int> parent(n, -1);
        while (!st.empty()) {
            int u = st.back();
            if (it[u] < (int)children[u].size()) {
                int v = children[u][it[u]++];
                parent[v] = u;
                st.push_back(v);
            } else {
                order.push_back(u);
                st.pop_back();
            }
        }

        auto mergeKnapsack = [&](const vector<long long>& A, const vector<long long>& B) {
            vector<long long> C(budget + 1, NEG);
            for (int i = 0; i <= budget; i++) if (A[i] != NEG) {
                for (int j = 0; j + i <= budget; j++) if (B[j] != NEG) {
                    C[i + j] = max(C[i + j], A[i] + B[j]);
                }
            }
            return C;
        };

        for (int u : order) {
            // Combine children assuming u not bought (children parentBought = false)
            vector<long long> comb0(budget + 1, NEG);
            comb0[0] = 0;
            for (int v : children[u]) {
                comb0 = mergeKnapsack(comb0, dp0[v]);
            }

            // Combine children assuming u bought (children parentBought = true)
            vector<long long> comb1(budget + 1, NEG);
            comb1[0] = 0;
            for (int v : children[u]) {
                comb1 = mergeKnapsack(comb1, dp1[v]);
            }

            // dp0[u]: parent not bought => u cost is full
            // Option not buy: comb0
            dp0[u] = comb0;
            // Option buy: pay full cost
            {
                int cost = present[u];
                long long prof = (long long)future[u] - cost;
                if (cost <= budget) {
                    for (int b = 0; b + cost <= budget; b++) if (comb1[b] != NEG) {
                        dp0[u][b + cost] = max(dp0[u][b + cost], comb1[b] + prof);
                    }
                }
            }

            // dp1[u]: parent bought => u cost is discounted
            dp1[u] = comb0;
            {
                int cost = present[u] / 2;
                long long prof = (long long)future[u] - cost;
                if (cost <= budget) {
                    for (int b = 0; b + cost <= budget; b++) if (comb1[b] != NEG) {
                        dp1[u][b + cost] = max(dp1[u][b + cost], comb1[b] + prof);
                    }
                }
            }
        }

        long long ans = 0;
        for (int b = 0; b <= budget; b++) ans = max(ans, dp0[0][b]);
        return (int)ans;
    }
};
// @lc code=end
